### PR TITLE
SegRep with Remote: Update components of segrep backpressure to suppo…

### PR DESCRIPTION
**Pull request by ankitkala**
…rt remote store.

### Description
Update components of segrep backpressure to support remote store.

### Related Issues
https://github.com/opensearch-project/OpenSearch/issues/7410

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: test/framework/src/main/java/org/opensearch/index/shard/IndexShardTestCase.java [hunk 1]
```diff
@@ -1302,15 +1302,18 @@ public final SegmentReplicationTargetService prepareForReplication(
         IndexShard primaryShard,
         IndexShard target,
         TransportService transportService,
-        IndicesService indicesService
+
```
Comment (scrawfor99) [inline test/framework/src/main/java/org/opensearch/index/shard/IndexShardTestCase.java:1305]: Nit: Extra line?

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java [hunk 2]
```diff
@@ -63,6 +66,8 @@ public static class Actions {
 
         public static final String GET_CHECKPOINT_INFO = "internal:index/shard/replication/get_checkpoint_info";
         public static final String GET_SEGMENT_FILES = "internal:index/shard/replication/get_segment_files";
+
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java]: ack

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java [hunk 3]
```diff
@@ -142,6 +162,20 @@ public void messageReceived(GetSegmentFilesRequest request, TransportChannel cha
         }
     }
 
+    private class UpdateVisibleCheckpointRequestHandler implements TransportRequestHandler<UpdateVisibleCheckpointRequest> {
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java:153]: I am following the pattern we're using for transport action & handlers in `SegmentReplicationSourceService`

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java [hunk 4]
```diff
@@ -54,6 +55,8 @@ public class SegmentReplicationSourceService extends AbstractLifecycleComponent
     private final TransportService transportService;
     private final IndicesService indicesService;
 
+    private final ClusterService clusterService;
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java]: My bad. It was part of the code which i eventually removed. Will update the PR before publishing

File: test/framework/src/main/java/org/opensearch/index/shard/IndexShardTestCase.java (hunk 1 above)
Comment (ankitkala) [reply test/framework/src/main/java/org/opensearch/index/shard/IndexShardTestCase.java]: done

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java [hunk 5]
```diff
@@ -274,6 +291,62 @@ public void onReplicationFailure(
...
+
+        final UpdateVisibleCheckpointRequest request = new UpdateVisibleCheckpointRequest(
+            replicationId,
+            replicaShard.routingEntry().allocationId().getId(),
+            primaryShard.shardId(),
+            getPrimaryNode(primaryShard),
+            replicaShard.getLatestReplicationCheckpoint()
+        );
+
+        final TransportRequestOptions options = TransportRequestOptions.builder()
```
Comment (mch2) [inline server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:305]: I'm wondering if we should move all this client code to PrimaryShardReplicationSource and fetch it from SegmentReplicationSourceFactory.  With the implementation of remote store integration the need for the ReplicationSource interface is gone, so we could do this in [refactoring](https://github.com/opensearch-project/OpenSearch/issues/7802) issue.

File: server/src/internalClusterTest/java/org/opensearch/index/SegmentReplicationPressureIT.java [hunk 6]
```diff
@@ -313,14 +318,23 @@ public void testBulkWritesRejected() throws Exception {
...
-        waitForSearchableDocs(totalDocs, replicaNodes.toArray(new String[] {}));
+        assertReplicaCheckpointUpdated(primaryShard);
 
         // index another doc showing there is no pressure enforced.
         executeBulkRequest(nodes, totalDocs);
         waitForSearchableDocs(totalDocs * 2L, replicaNodes.toArray(new String[] {}));
         verifyStoreContent();
     }
 
+    private void assertReplicaCheckpointUpdated(IndexShard primaryShard) throws Exception {
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply server/src/internalClusterTest/java/org/opensearch/index/SegmentReplicationPressureIT.java:329]: Ack. I'll move it to `SegmentReplicationBaseIT`.

Nice catch around the replica count. I'll add this a validation as well.

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java [hunk 7]
```diff
@@ -880,7 +877,7 @@ public void testPressureServiceStats() throws Exception {
                 assertEquals(1, shardStatsSet.size());
                 final SegmentReplicationShardStats stats = shardStatsSet.stream().findFirst().get();
                 assertEquals(0, stats.getCheckpointsBehindCount());
-            });
+            }, 30, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:880]: I had a test run failing here so i increased the timeout. Let me try reverting this change and test. If the test is passing consistently, i'll remove this.

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java [hunk 8]
```diff
@@ -274,6 +291,62 @@ public void onReplicationFailure(
...
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error(
+                    "Failed to update visible checkpoint for replica {}, {}: {}",
+                    replicaShard.shardId(),
+                    request.getCheckpoint(),
+                    e
+                );
+                transportClient.cancel();
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:333]: Sure. Removed the cancel logic.

File: server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java (hunk 5 above)
Comment (ankitkala) [reply server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java:305]: Sure.
 I originally started with the same thought but then realised that i'll have add this logic in 2 places `PrimaryShardReplicationSource` and `RemoteStoreReplicationSource`(or make the `SegmentReplicationSource` abstract). Thus decided against it. 
We can address this with refactor.

File: server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java [hunk 9]
```diff
@@ -1174,13 +1174,18 @@ public synchronized void updateVisibleCheckpointForShard(final String allocation
...
+            logger.warn("Ignoring the checkpoint update for allocation ID {} as its not being tracked by primary", allocationId);
+            return;
+        }
         if (cps.checkpointTimers.isEmpty() == false) {
             // stop any timers for checkpoints up to the received cp and remove from cps.checkpointTimers.
             // Compute the max lag from the set of completed timers.
             final AtomicLong lastFinished = new AtomicLong(0L);
             cps.checkpointTimers.entrySet().removeIf((entry) -> {
-                boolean result = visibleCheckpoint.equals(entry.getKey()) || visibleCheckpoint.isAheadOf(entry.getKey());
+                boolean result = !entry.getKey().isAheadOf(visibleCheckpoint);
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java:1188]: Addressed in the latest revision.

**Review by mch2 — APPROVED** [review]

**Pull request merged**